src/main/java/org/ethereum/vm/DataWord.java

```diff
@@ -0,0 +1,344 @@
+/*
+ * Copyright (c) [2018] [ The Semux Developers ]
+ * Copyright (c) [2016] [ <ether.camp> ]
+ * This file is part of the ethereumJ library.
+ *
+ * The ethereumJ library is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The ethereumJ library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with the ethereumJ library. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.ethereum.vm;
+
+import java.math.BigInteger;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+
+import org.ethereum.vm.util.ByteArrayUtil;
+import org.ethereum.vm.util.HexUtil;
+
+/**
+ * DataWord is the 32-byte array representation of a 256-bit number.
+ *
+ * @ImplNote DataWord objects are immutable.
+ */
+public class DataWord implements Comparable<DataWord> {
+
+    public static final BigInteger TWO_POW_256 = BigInteger.valueOf(2).pow(256);
+    public static final BigInteger MAX_VALUE = TWO_POW_256.subtract(BigInteger.ONE);
+
+    public static final DataWord ZERO = new DataWord(0);
+    public static final DataWord ONE = new DataWord(1);
+
+    public static final int SIZE = 32;
+
+    private byte[] data = new byte[SIZE];
+
+    public DataWord(int num) {
+        this(ByteBuffer.allocate(4).putInt(num).array());
+    }
+
+    public DataWord(long num) {
+        this(ByteBuffer.allocate(8).putLong(num).array());
+    }
+
+    public DataWord(BigInteger num) {
+        // NOTE: the MAX_VALUE will produce a 33 bytes array
+        byte[] arr = num.toByteArray();
+        int offset = Math.max(arr.length - 32, 0);
+        int size = arr.length - offset;
+        System.arraycopy(arr, offset, data, SIZE - size, size);
+    }
+
+    public DataWord(String hex) {
+        this(HexUtil.fromHexString(hex));
+    }
+
+    public DataWord(byte[] data) {
+        if (data.length == 32) {
+            this.data = data.clone();
+        } else if (data.length < 32) {
+            System.arraycopy(data, 0, this.data, 32 - data.length, data.length);
+        } else {
+            throw new RuntimeException("Data word can't exceed 32 bytes");
+        }
+    }
+
+    /**
+     * Returns a clone of the underlying byte array.
+     *
+     * @return a byte array
+     */
+    public byte[] getData() {
+        return data.clone();
+    }
+
+    public byte[] getLast20Bytes() {
+        return Arrays.copyOfRange(data, data.length - 20, data.length);
+    }
+
+    public byte getByte(int index) {
+        if (index < 0 || index >= SIZE) {
```

Comment:
Please correct me if I am wrong, but ain't the bound checks embedded within Java arrays? I thought it's the primary reason why the arrays are slow – because one cannot disable the bounds check.